File: python/deltarobot/trajectory.py
"""Time scaling profiles and paths (Cartesian line, pick-and-place arch, joint space).

A *profile* maps time to distance along a path: s(t), s'(t), s''(t) for 0 <= t <= T.
Profiles: "trapezoid", "scurve" (sin^2 acceleration), "quintic", "cycloidal".
Mirrored in assets/js/delta/trajectory.js.
"""

from __future__ import annotations

import math
from typing import List, Sequence, Tuple

Vec = Tuple[float, float, float]
PROFILES = ("trapezoid", "scurve", "quintic", "cycloidal")


class Profile:
    """Rest-to-rest time scaling over distance d with speed/accel limits."""

    def __init__(self, kind: str, d: float, vmax: float, amax: float):
        if kind not in PROFILES:
            raise ValueError("profile must be one of %s" % (PROFILES,))
        if vmax <= 0 or amax <= 0:
            raise ValueError("vmax and amax must be positive")
        self.kind, self.d, self.vmax, self.amax = kind, max(0.0, d), vmax, amax
        d = self.d
        if d == 0:
            self.T = 0.0
            return
        if kind == "trapezoid":
            ta = vmax / amax
            if amax * ta * ta >= d:           # triangular
                ta = math.sqrt(d / amax)
                self.vp = amax * ta
                self.ta, self.tc = ta, 0.0
            else:
                self.vp = vmax
                self.ta, self.tc = ta, (d - vmax * ta) / vmax
            self.T = 2 * self.ta + self.tc
        elif kind == "scurve":
            # a(t) = A sin^2(pi t / ta): mean accel A/2, reaches v = A ta / 2
            A = amax
            ta = 2 * vmax / A
            if A * ta * ta / 2 >= d:          # distance of accel+decel phases = v * ta
                ta = math.sqrt(2 * d / A)
                self.vp = A * ta / 2
                self.ta, self.tc = ta, 0.0
            else:
                self.vp = vmax
                self.ta, self.tc = ta, (d - vmax * ta) / vmax
            self.A = A
            self.T = 2 * self.ta + self.tc
        elif kind == "quintic":
            self.T = max(15 * d / (8 * vmax), math.sqrt(10 * d / (math.sqrt(3) * amax)))
        else:  # cycloidal
            self.T = max(2 * d / vmax, math.sqrt(2 * math.pi * d / amax))

    def at(self, t: float) -> Tuple[float, float, float]:
        """(s, s_dot, s_ddot) at time t (clamped to [0, T])."""
        d, T = self.d, self.T
        if T <= 0:
            return 0.0, 0.0, 0.0
        t = min(max(t, 0.0), T)
        k = self.kind
        if k == "trapezoid":
            ta, tc, vp = self.ta, self.tc, self.vp
            a = vp / ta
            if t < ta:
                return 0.5 * a * t * t, a * t, a
            if t < ta + tc:
                return 0.5 * a * ta * ta + vp * (t - ta), vp, 0.0
            u = T - t
            return d - 0.5 * a * u * u, a * u, -a if t < T else 0.0
        if k == "scurve":
            ta, tc, vp, A = self.ta, self.tc, self.vp, self.A
            A = 2 * vp / ta                     # actual peak accel for this move

            def acc_phase(u: float) -> Tuple[float, float, float]:
                w = math.pi / ta
                s = A * (u * u / 4 - (1 - math.cos(2 * w * u)) / (8 * w * w))
                v = A * (u / 2 - math.sin(2 * w * u) / (4 * w))
                a = A * math.sin(w * u) ** 2
                return s, v, a

            if t < ta:
                return acc_phase(t)
            s_a = acc_phase(ta)[0]
            if t < ta + tc:
                return s_a + vp * (t - ta), vp, 0.0
            s, v, a = acc_phase(T - t)
            return d - s, v, -a
        if k == "quintic":
            x = t / T
            s = d * (10 * x ** 3 - 15 * x ** 4 + 6 * x ** 5)
            v = d * (30 * x ** 2 - 60 * x ** 3 + 30 * x ** 4) / T
            a = d * (60 * x - 180 * x ** 2 + 120 * x ** 3) / (T * T)
            return s, v, a
        x = t / T
        s = d * (x - math.sin(2 * math.pi * x) / (2 * math.pi))
        v = d * (1 - math.cos(2 * math.pi * x)) / T
        a = d * 2 * math.pi * math.sin(2 * math.pi * x) / (T * T)
        return s, v, a


# ---------------------------------------------------------------- paths
def _lerp(a: Sequence[float], b: Sequence[float], u: float) -> Vec:
    return (a[0] + (b[0] - a[0]) * u, a[1] + (b[1] - a[1]) * u, a[2] + (b[2] - a[2]) * u)


def _dist(a: Sequence[float], b: Sequence[float]) -> float:
    return math.sqrt(sum((a[i] - b[i]) ** 2 for i in range(3)))


class Path:
    """Piecewise path made of straight lines and circular arcs, parametrised by arc length."""

    def __init__(self) -> None:
        self.segs: List[tuple] = []   # ("line", p0, p1, len) | ("arc", centre, u, w, radius, angle, len)
        self.length = 0.0

    def line(self, p0: Sequence[float], p1: Sequence[float]) -> "Path":
        n = _dist(p0, p1)
        if n > 1e-12:
            self.segs.append(("line", tuple(p0), tuple(p1), n))
            self.length += n
        return self

    def arc(self, centre: Sequence[float], u: Sequence[float], w: Sequence[float],
            radius: float, angle: float) -> "Path":
        """Arc: centre + radius*(cos(a) u + sin(a) w), a from 0 to angle."""
        n = radius * abs(angle)
        if n > 1e-12:
            self.segs.append(("arc", tuple(centre), tuple(u), tuple(w), radius, angle, n))
            self.length += n
        return self

    def point(self, s: float) -> Vec:
        s = min(max(s, 0.0), self.length)
        for seg in self.segs:
            n = seg[-1]
            if s <= n or seg is self.segs[-1]:
                if seg[0] == "line":
                    return _lerp(seg[1], seg[2], s / n if n else 0.0)
                _, c, u, w, rad, ang, _n = seg
                a = ang * (s / n)
                ca, sa = math.cos(a), math.sin(a)
                return (c[0] + rad * (ca * u[0] + sa * w[0]),
                        c[1] + rad * (ca * u[1] + sa * w[1]),
                        c[2] + rad * (ca * u[2] + sa * w[2]))
            s -= n
        return tuple(self.segs[-1][2]) if self.segs else (0.0, 0.0, 0.0)  # type: ignore[return-value]


def line_path(p0: Sequence[float], p1: Sequence[float]) -> Path:
    return Path().line(p0, p1)


def arch_path(p0: Sequence[float], p1: Sequence[float], height: float, radius: float = 0.0) -> Path:
    """Pick-and-place 'door' path: up from p0, across, down to p1, rounded corners.

    The travel height is max(p0.z, p1.z) + height. radius defaults to height/2.
    """
    ztop = max(p0[2], p1[2]) + height
    a = (p0[0], p0[1], ztop)
    b = (p1[0], p1[1], ztop)
    horiz = _dist(a, b)
    path = Path()
    if horiz < 1e-9:
        return path.line(p0, a).line(a, p1) if height > 0 else path.line(p0, p1)
    rad = radius or height / 2.0
    rad = max(0.0, min(rad, ztop - p0[2], ztop - p1[2], horiz / 2.0))
    hdir = ((b[0] - a[0]) / horiz, (b[1] - a[1]) / horiz, 0.0)
    up = (0.0, 0.0, 1.0)
    # corner 1 at a: coming up, leaving along hdir
    c1 = (a[0] + hdir[0] * rad, a[1] + hdir[1] * rad, ztop - rad)
    path.line(p0, (a[0], a[1], ztop - rad))
    if rad > 0:
        path.arc(c1, (-hdir[0], -hdir[1], 0.0), up, rad, math.pi / 2)
    path.line((a[0] + hdir[0] * rad, a[1] + hdir[1] * rad, ztop),
              (b[0] - hdir[0] * rad, b[1] - hdir[1] * rad, ztop))
    c2 = (b[0] - hdir[0] * rad, b[1] - hdir[1] * rad, ztop - rad)
    if rad > 0:
        path.arc(c2, up, hdir, rad, math.pi / 2)
    path.line((b[0], b[1], ztop - rad), p1)
    return path


def sample_path(path: Path, profile_kind: str, vmax: float, amax: float, dt: float,
                t0: float = 0.0) -> List[Tuple[float, Vec]]:
    """Sample (t, point) along a path; always includes the final point."""
    prof = Profile(profile_kind, path.length, vmax, amax)
    out: List[Tuple[float, Vec]] = []
    n = max(1, int(math.ceil(prof.T / dt)))
    for k in range(1, n + 1):
        t = min(prof.T, k * dt)
        out.append((t0 + t, path.point(prof.at(t)[0])))
    return out


def sample_joint(q0: Sequence[float], q1: Sequence[float], profile_kind: str, wmax: float,
                 alpha: float, dt: float, t0: float = 0.0) -> List[Tuple[float, Vec]]:
    """Synchronised joint-space move: the largest joint move sets the timing."""
    dq = [q1[i] - q0[i] for i in range(3)]
    big = max(abs(v) for v in dq)
    prof = Profile(profile_kind, big, wmax, alpha)
    out: List[Tuple[float, Vec]] = []
    n = max(1, int(math.ceil(prof.T / dt)))
    for k in range(1, n + 1):
        t = min(prof.T, k * dt)
        u = prof.at(t)[0] / big if big > 0 else 1.0
        out.append((t0 + t, (q0[0] + dq[0] * u, q0[1] + dq[1] * u, q0[2] + dq[2] * u)))
    return out

File: assets/js/delta/trajectory.js
// trajectory.js — time scaling profiles and paths. Mirror of python/deltarobot/trajectory.py.

export const PROFILES = ['trapezoid', 'scurve', 'quintic', 'cycloidal'];
export const PROFILE_KO = { trapezoid: '사다리꼴', scurve: 'S-커브', quintic: '5차 다항식', cycloidal: '사이클로이드' };

export class Profile {
  constructor(kind, d, vmax, amax) {
    if (!PROFILES.includes(kind)) throw new Error('profile must be one of ' + PROFILES.join(', '));
    if (!(vmax > 0) || !(amax > 0)) throw new Error('vmax and amax must be positive');
    this.kind = kind; this.d = Math.max(0, d); this.vmax = vmax; this.amax = amax;
    d = this.d;
    if (d === 0) { this.T = 0; return; }
    if (kind === 'trapezoid') {
      let ta = vmax / amax;
      if (amax * ta * ta >= d) { ta = Math.sqrt(d / amax); this.vp = amax * ta; this.ta = ta; this.tc = 0; }
      else { this.vp = vmax; this.ta = ta; this.tc = (d - vmax * ta) / vmax; }
      this.T = 2 * this.ta + this.tc;
    } else if (kind === 'scurve') {
      const A = amax;
      let ta = 2 * vmax / A;
      if (A * ta * ta / 2 >= d) { ta = Math.sqrt(2 * d / A); this.vp = A * ta / 2; this.ta = ta; this.tc = 0; }
      else { this.vp = vmax; this.ta = ta; this.tc = (d - vmax * ta) / vmax; }
      this.A = A;
      this.T = 2 * this.ta + this.tc;
    } else if (kind === 'quintic') {
      this.T = Math.max(15 * d / (8 * vmax), Math.sqrt(10 * d / (Math.sqrt(3) * amax)));
    } else {
      this.T = Math.max(2 * d / vmax, Math.sqrt(2 * Math.PI * d / amax));
    }
  }

  at(t) {
    const d = this.d, T = this.T;
    if (T <= 0) return [0, 0, 0];
    t = Math.min(Math.max(t, 0), T);
    const k = this.kind;
    if (k === 'trapezoid') {
      const { ta, tc, vp } = this;
      const a = vp / ta;
      if (t < ta) return [0.5 * a * t * t, a * t, a];
      if (t < ta + tc) return [0.5 * a * ta * ta + vp * (t - ta), vp, 0];
      const u = T - t;
      return [d - 0.5 * a * u * u, a * u, t < T ? -a : 0];
    }
    if (k === 'scurve') {
      const { ta, tc, vp } = this;
      const A = 2 * vp / ta;
      const w = Math.PI / ta;
      const acc = (u) => [
        A * (u * u / 4 - (1 - Math.cos(2 * w * u)) / (8 * w * w)),
        A * (u / 2 - Math.sin(2 * w * u) / (4 * w)),
        A * Math.sin(w * u) ** 2,
      ];
      if (t < ta) return acc(t);
      const sa = acc(ta)[0];
      if (t < ta + tc) return [sa + vp * (t - ta), vp, 0];
      const [s, v, a] = acc(T - t);
      return [d - s, v, -a];
    }
    const x = t / T;
    if (k === 'quintic') {
      return [d * (10 * x ** 3 - 15 * x ** 4 + 6 * x ** 5),
        d * (30 * x ** 2 - 60 * x ** 3 + 30 * x ** 4) / T,
        d * (60 * x - 180 * x ** 2 + 120 * x ** 3) / (T * T)];
    }
    return [d * (x - Math.sin(2 * Math.PI * x) / (2 * Math.PI)),
      d * (1 - Math.cos(2 * Math.PI * x)) / T,
      d * 2 * Math.PI * Math.sin(2 * Math.PI * x) / (T * T)];
  }
}

const lerp = (a, b, u) => [a[0] + (b[0] - a[0]) * u, a[1] + (b[1] - a[1]) * u, a[2] + (b[2] - a[2]) * u];
const dist = (a, b) => Math.hypot(a[0] - b[0], a[1] - b[1], a[2] - b[2]);

export class Path {
  constructor() { this.segs = []; this.length = 0; }
  line(p0, p1) {
    const n = dist(p0, p1);
    if (n > 1e-12) { this.segs.push({ type: 'line', p0: p0.slice(), p1: p1.slice(), len: n }); this.length += n; }
    return this;
  }
  arc(c, u, w, radius, angle) {
    const n = radius * Math.abs(angle);
    if (n > 1e-12) { this.segs.push({ type: 'arc', c, u, w, radius, angle, len: n }); this.length += n; }
    return this;
  }
  point(s) {
    s = Math.min(Math.max(s, 0), this.length);
    for (let i = 0; i < this.segs.length; i++) {
      const g = this.segs[i];
      if (s <= g.len || i === this.segs.length - 1) {
        if (g.type === 'line') return lerp(g.p0, g.p1, g.len ? s / g.len : 0);
        const a = g.angle * (s / g.len), ca = Math.cos(a), sa = Math.sin(a);
        return [0, 1, 2].map((k) => g.c[k] + g.radius * (ca * g.u[k] + sa * g.w[k]));
      }
      s -= g.len;
    }
    return [0, 0, 0];
  }
}

export const linePath = (p0, p1) => new Path().line(p0, p1);

export function archPath(p0, p1, height, radius = 0) {
  const ztop = Math.max(p0[2], p1[2]) + height;
  const a = [p0[0], p0[1], ztop], b = [p1[0], p1[1], ztop];
  const horiz = dist(a, b);
  const path = new Path();
  if (horiz < 1e-9) return height > 0 ? path.line(p0, a).line(a, p1) : path.line(p0, p1);
  let rad = radius || height / 2;
  rad = Math.max(0, Math.min(rad, ztop - p0[2], ztop - p1[2], horiz / 2));
  const h = [(b[0] - a[0]) / horiz, (b[1] - a[1]) / horiz, 0];
  const up = [0, 0, 1];
  path.line(p0, [a[0], a[1], ztop - rad]);
  if (rad > 0) path.arc([a[0] + h[0] * rad, a[1] + h[1] * rad, ztop - rad], [-h[0], -h[1], 0], up, rad, Math.PI / 2);
  path.line([a[0] + h[0] * rad, a[1] + h[1] * rad, ztop], [b[0] - h[0] * rad, b[1] - h[1] * rad, ztop]);
  if (rad > 0) path.arc([b[0] - h[0] * rad, b[1] - h[1] * rad, ztop - rad], up, h, rad, Math.PI / 2);
  path.line([b[0], b[1], ztop - rad], p1);
  return path;
}

export function samplePath(path, kind, vmax, amax, dt, t0 = 0) {
  const prof = new Profile(kind, path.length, vmax, amax);
  const out = [];
  const n = Math.max(1, Math.ceil(prof.T / dt));
  for (let k = 1; k <= n; k++) {
    const t = Math.min(prof.T, k * dt);
    out.push([t0 + t, path.point(prof.at(t)[0])]);
  }
  return out;
}

export function sampleJoint(q0, q1, kind, wmax, alpha, dt, t0 = 0) {
  const dq = [0, 1, 2].map((i) => q1[i] - q0[i]);
  const big = Math.max(...dq.map(Math.abs));
  const prof = new Profile(kind, big, wmax, alpha);
  const out = [];
  const n = Math.max(1, Math.ceil(prof.T / dt));
  for (let k = 1; k <= n; k++) {
    const t = Math.min(prof.T, k * dt);
    const u = big > 0 ? prof.at(t)[0] / big : 1;
    out.push([t0 + t, [0, 1, 2].map((i) => q0[i] + dq[i] * u)]);
  }
  return out;
}
